### DISCUSSION TITLE
How to find exactly which lines are being tree-shaken?

### DISCUSSION THREAD

**Discussion by hdragos** [webpack/webpack#19059] [Q&A] [answered]
We work with a large and dynamic codebase that gets a lot of features added. Sometimes these features are disabled and forgotten in the codebase. People come and go, priorities shift and sometimes we end up with dead code that nobody knows about.

I know that dead code files can be easily identified using a plugin such as [unused-files-webpack-plugin](https://www.npmjs.com/package/unused-files-webpack-plugin), however I would say that most dead code is found within used files (basically unused methods of used classes / modules).

Is there a way to get a full list of lines / methods / classes / files that were tree shaken? We want to do this in order to see exactly what dead code we have in our codebase.

**We use Webpack5 for building our frontend.**

**Comment by alexander-akait** [comment, marked as answer]
Unfortunately there is no such possibility

**Reply by hdragos** [reply]
Well, in that case there are only 2 possibilities left:
1. Write a plugin that does that
2. Fork the repo and try to implement it directly into webpack

Is there a way to hook into the tree-shaking process? I would go for a building a plugin rather than making a feature that would rarely be used. If not, where is the tree-shaking process taking place? Just a file / function name would suffice, I will dig deeper from there.

**Reply by alexander-akait** [reply]
> Is there a way to hook into the tree-shaking process?

There are many different places and many plugins, which do diffrent things, there is no single place where it is possible to implement. 

sideEffect, dead code, minimizer, unused exports, pure notatitions, there are many tricks for code optimization

**Reply by alexander-akait** [reply]
You can try to disable all optimizations and make a build, then enable them all, except for deleting space and possibly other options in the minimizer and compare the files

**Comment by gouravchahar13** [comment]
Hello sir,
I am new here so if I say something stupid please correct me but  I believe integrating code coverage tools like Istanbul or NYC could help identify unused code and give us insights into the tree-shaking process what are your thoughts on this? @alexander-akait
